[LoopVectorize] Refine runtime memory check costs when there is an outer loop

david-arm opened the pull request:
When we generate runtime memory checks for an inner loop it's
possible that these checks are invariant in the outer loop and
so will get hoisted out. In such cases, the effective cost of
the checks should reduce to reflect the outer loop trip count.

This fixes a 25% performance regression introduced by commit

49b0e6dcc296792b577ae8f0f674e61a0929b99d

when building the SPEC2017 x264 benchmark with PGO, where we
decided the inner loop trip count wasn't high enough to warrant
the (incorrect) high cost of the runtime checks. Also, when
runtime memory checks consist entirely of diff checks these are
likely to be outer loop invariant.

david-arm committed:
[LoopVectorize] Refine runtime memory check costs when there is an outer loop

When we generate runtime memory checks for an inner loop it's
possible that these checks are invariant in the outer loop and
so will get hoisted out. In such cases, the effective cost of
the checks should reduce to reflect the outer loop trip count.

This fixes a 25% performance regression introduced by commit

49b0e6dcc296792b577ae8f0f674e61a0929b99d

when building the SPEC2017 x264 benchmark with PGO, where we
decided the inner loop trip count wasn't high enough to warrant
the (incorrect) high cost of the runtime checks. Also, when
runtime memory checks consist entirely of diff checks these are
likely to be outer loop invariant.

david-arm committed:
Add tests showing runtime checks cost with low trip counts

Rin18 commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_memcheck_cost.ll:
```diff
@@ -0,0 +1,193 @@
+target triple = "aarch64-unknown-linux-gnu"
+
+
+define void @outer_no_tc(ptr nocapture noundef %a, ptr nocapture noundef readonly %b, i64 noundef %m, i64 noundef %n) vscale_range(1,16) {
```
Is the `vscale_range(1, 16)` needed here (and for the other tests)? I don't think it makes any difference to the output.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_memcheck_cost.ll:
```diff
@@ -0,0 +1,193 @@
+; RUN: opt -p loop-vectorize -mattr=+sve -debug-only=loop-vectorize -S -disable-output < %s 2>&1 | FileCheck %s
+
```
This needs `REQUIRES: asserts` as this is checking the debug output.

davemgreen commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_memcheck_cost.ll:
```diff
@@ -0,0 +1,187 @@
+target triple = "aarch64-unknown-linux-gnu"
+
+
+define void @outer_no_tc(ptr nocapture noundef %a, ptr nocapture noundef readonly %b, i64 noundef %m, i64 noundef %n) {
```
It might be worth adding a version with no outer loop as a baseline. And maybe a comment about what the test is testing.

davemgreen reviewed (approved):
This seems OK from what I can tell. LGTM

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2091,16 +2091,45 @@ class GeneratedRTChecks {
+        ScalarEvolution *SE = MemCheckExp.getSE();
+        const SCEV *Cond = SE->getSCEV(MemRuntimeCheckCond);
+        if (SE->isLoopInvariant(Cond, OuterLoop)) {
+          if (std::optional<unsigned> OuterTC =
```
would it be simpler/more compact to use `getSmallBestKnownTC(*SE, OuterLoop).value_or(2)` here and

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2091,16 +2091,45 @@ class GeneratedRTChecks {
+          }
+
+          // Let's ensure the cost is always at least 1.
+          if (MemCheckCost == 0)
```
more compact to use `std::max` possibly when adding to `RTCheckCost`?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2091,16 +2091,45 @@ class GeneratedRTChecks {
+      // according to the outer loop trip count.
+      if (OuterLoop) {
+        ScalarEvolution *SE = MemCheckExp.getSE();
+        const SCEV *Cond = SE->getSCEV(MemRuntimeCheckCond);
```
Is it possible to have a mixture of invariant/variant checks or at least some invariant instructions generated? If so, it would probably be more accurate to check for each generated instruction?

david-arm replied:
So I did think about that, but I chose not to do that for a couple of reasons:

1. I assume that getting SCEVs is quite expensive, unless they've already been cached before. So I was worried about the increase in compilation time by checking each instruction.
2. Even if a particular instruction in the sequence is invariant, there is no guarantee it will be hoisted if the use is not invariant for some reason. So I thought the most convincing case was when the final condition was invariant as that likely means the whole sequence will be hoisted.

I'm happy to refine this algorithm further if you still think it's worth the extra compilation time and the risk of under-estimating the cost for instructions that don't get hoisted?

fhahn replied:
> I assume that getting SCEVs is quite expensive, unless they've already been cached before. So I was worried about the increase in compilation time by checking each instruction.

I think `SE->getSCEV(MemRuntimeCheckCond)` will cause SCEV expressions to be built anyways for all the instructions feeding `MemRuntimeCheckCond`, so the time spent in SCEV construction should be roughly the same.

> Even if a particular instruction in the sequence is invariant, there is no guarantee it will be hoisted if the use is not invariant for some reason. So I thought the most convincing case was when the final condition was invariant as that likely means the whole sequence will be hoisted.

I think LLVM's LICM will hoist any hoist able instruction, the backend may sink them back. Loop-invariant instructions may also be sunk back in the loop (e.g. if it reduces register pressure or something like that), so at this point there's no hard guarantees I think. 

it might not be worth including this in the initial version, but it would at least be good to include a comment on why we only the final condition is checked.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2070,7 +2070,7 @@ class GeneratedRTChecks {
   }
 
-  InstructionCost getCost() {
+  InstructionCost getCost(Loop *OuterLoop) {
```
It might be better to store the inner or outer loop directly in the struct; the inner loop is passed in to `Create` and there are 2 places that check for an outer loop already (via `if (auto *PL = LI->getLoopFor(LoopVectorPreHeader))`)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2091,16 +2091,45 @@ class GeneratedRTChecks {
+        const SCEV *Cond = SE->getSCEV(MemRuntimeCheckCond);
+        if (SE->isLoopInvariant(Cond, OuterLoop)) {
+          if (std::optional<unsigned> OuterTC =
+                  getSmallBestKnownTC(*SE, OuterLoop))
```
Hmm, not sure if this is the right function to use here, as it will use the maximum trip count if no info is present.

For something like  below, it will divide by something like 2^32, so it is likely to over-estimate real trip counts in a lot of cases by quite a large margin

```
define void @outer_no_tc(ptr nocapture noundef %a, ptr nocapture noundef readonly %b, i32 noundef %m, i64 noundef %n) {
entry:
  %e = zext i32 %m to i64
  %cmp = icmp ugt i64 %e, 0
   br i1 %cmp, label %outer.loop, label %outer.exit

 outer.loop:
   %outer.iv = phi i64 [ %outer.iv.next, %inner.exit ], [ 0, %entry ]
   %mul.us = mul nsw i64 %outer.iv, %n
   br label %inner.loop

 inner.loop:
   %inner.iv = phi i64 [ 0, %outer.loop ], [ %inner.iv.next, %inner.loop ]
   %add.us = add nuw nsw i64 %inner.iv, %mul.us
   %arrayidx.us = getelementptr inbounds i8, ptr %b, i64 %add.us
   %0 = load i8, ptr %arrayidx.us, align 1
   %arrayidx7.us = getelementptr inbounds i8, ptr %a, i64 %add.us
   %1 = load i8, ptr %arrayidx7.us, align 1
   %add9.us = add i8 %1, %0
   store i8 %add9.us, ptr %arrayidx7.us, align 1
   %inner.iv.next = add nuw nsw i64 %inner.iv, 1
   %exitcond.not = icmp eq i64 %inner.iv.next, %n
   br i1 %exitcond.not, label %inner.exit, label %inner.loop

 inner.exit:
   %outer.iv.next = add nuw nsw i64 %outer.iv, 1
   %exitcond27.not = icmp eq i64 %outer.iv.next, %e
   br i1 %exitcond27.not, label %outer.exit, label %outer.loop

 outer.exit:
   ret void
 }
```

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2091,16 +2091,45 @@ class GeneratedRTChecks {
+    }
+
     if (SCEVCheckBlock || MemCheckBlock)
       LLVM_DEBUG(dbgs() << "Total cost of runtime checks: " << RTCheckCost
```
We should probably mention the division of the mem-check cost somewhere in the debug output, as. currently it could be confusing to see the costs printed per instructions not matching up with the computed cost.

david-arm committed:
Revert "[LoopVectorize] Refine runtime memory check costs when there is an outer loop"

This reverts commit a152314dcca6e0210ed747432fa1aed1c3dcfc10.

david-arm committed:
Revert "Add tests showing runtime checks cost with low trip counts"

This reverts commit a4caa47dc8d2db75f6bb2ac3f880da4e1f6bea82.

david-arm committed:
Add tests showing runtime checks cost with low trip counts

david-arm committed:
[LoopVectorize] Refine runtime memory check costs when there is an outer loop

When we generate runtime memory checks for an inner loop it's
possible that these checks are invariant in the outer loop and
so will get hoisted out. In such cases, the effective cost of
the checks should reduce to reflect the outer loop trip count.

This fixes a 25% performance regression introduced by commit

49b0e6dcc296792b577ae8f0f674e61a0929b99d

when building the SPEC2017 x264 benchmark with PGO, where we
decided the inner loop trip count wasn't high enough to warrant
the (incorrect) high cost of the runtime checks. Also, when
runtime memory checks consist entirely of diff checks these are
likely to be outer loop invariant.

david-arm commented:
Gentle ping!

Rin18 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2076,16 +2081,61 @@ class GeneratedRTChecks {
+          InstructionCost NewMemCheckCost = MemCheckCost / BestTripCount;
+
+          // Let's ensure the cost is always at least 1.
+          NewMemCheckCost = std::max(*NewMemCheckCost.getValue(), (long)1);
```
There's a buildbot failure at this line. Has that been fixed? Might be worth getting that triggered again.

david-arm replied:
Good spot! I hope I've fixed it now. :)

Asher8118 reviewed (commented):
One small comment, but otherwise LGTM! I'll leave someone else more familiar with the code to approve the change.

david-arm commented:
Trying a rebase to see if that fixes the failing RISCV test that appears unrelated to my change

fhahn reviewed (commented):
Thanks for the updates!

> > Trying a rebase to see if that fixes the failing RISCV test that appears unrelated to my change

Alternatively you could use the `Update Branch` button to merge in the changes from current main, which I think is preferred to rebases, as the GH UI may lose comments after rebases.

david-arm commented:
> Thanks for the updates!
> 
> > > Trying a rebase to see if that fixes the failing RISCV test that appears unrelated to my change
> 
> Alternatively you could use the `Update Branch` button to merge in the changes from current main, which I think is preferred to rebases, as the GH UI may lose comments after rebases.

Yeah I used to use the "Update Branch" button, but I find it makes life very difficult for reviewers (and authors) and I'd like to stop using it in future. The problem is that in Phabricator we could do stacked patches quite easily, i.e. NFC test patch followed by code change. The purpose of this is to allow people to see the impact of the code changes. In this new world we're supposed to always add a new commit every time we address comments, which completely defeats the whole purpose of splitting the test patch up from the code change. Quite often reviewers ask authors to rewrite tests or add new ones, which means that when you add a new commit you can no longer see the impact of your code change. It feels like we should either:

1. Stop splitting commits up into NFC test + code change patch, because it's not that useful anymore. Or
2. Force push the same two commits every time similar to how we did things on Phabricator. That way when we do change the tests it's still possible for the reviewer to see what effect the code change has.

To be honest either approach is unsatisfactory, but I feel option 2) gives the reviewer a more fighting chance of seeing cause and effect!

fhahn commented:
> To be honest either approach is unsatisfactory, but I feel option 2) gives the reviewer a more fighting chance of seeing cause and effect!

IMO including a separate commit for precommitting  tests in a feature PR isn't ideal, as the default view in GH shows the diff for all changes compared to the main branch and when merging the PR all commits get squashed into one.

It's now possible to get something close to Phabricator's stacked reviews discussed here https://discourse.llvm.org/t/update-on-github-pull-requests/71540/146.

david-arm commented:
> Could you update the places that look up the outer loop to use the new variable as well?

Hi @fhahn, I couldn't find any other instances in GeneratedRTChecks where we call `getParentLoop()`. I'm not sure where else I could use `OuterLoop`?

fhahn commented:
> > Could you update the places that look up the outer loop to use the new variable as well?
> 
> Hi @fhahn, I couldn't find any other instances in GeneratedRTChecks where we call `getParentLoop()`. I'm not sure where else I could use `OuterLoop`?

There are 2 places that do something like `if (auto *PL = LI->getLoopFor(LoopVectorPreHeader)...`, which effectively checks if there's an outer loop

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2076,16 +2081,62 @@ class GeneratedRTChecks {
+        if (SE->isLoopInvariant(Cond, OuterLoop)) {
+          // It seems reasonable to assume that we can reduce the effective
+          // cost of the checks even when we know nothing about the trip
+          // count. Here I've assumed that the outer loop executes at least
```
nit: still refers to `I`, better to rephrase or use more general `we`

david-arm replied:
Silly me. Done!

fhahn reviewed (approved):
LGTM, thanks!
